### PR TITLE
AUT-2789: Log user out when they enter their password incorrectly max…

### PR DESCRIPTION
**Pull request by andrew-moores**
… allowed times.

## What

When the user fails to present their correct password more than maximum number of allowed retries they will no longer have their account locked.  Instead the count of failed password attempts will be removed and the caller of the API, the frontend, will be expected to log the user out.

## How to review

1. Code Review
2. Review with matching frontend changes.

## Checklist

- [ ] Impact on orch and auth mutual dependencies has been checked.

- [ ] A UCD review has been performed.

### CODE REVIEW

**Review by mattvot — CHANGES_REQUESTED** [review]
Looking good. I've left some questions on the code. 

Could the commit messages here mention a bit more detail on why some of the changes have been made please? I don't think it's this code that actually does the "Log user out when" bit. Maybe mention/reference the changes on the frontend in the commit message.

Thanks!

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/LoginHandler.java [hunk 1]
```diff
@@ -58,7 +58,15 @@ public class LoginHandler extends BaseFrontendHandler<LoginRequest>
 
     public static final String CODE_BLOCKED_KEY_PREFIX =
             CodeStorageService.PASSWORD_BLOCKED_KEY_PREFIX + JourneyType.PASSWORD_RESET;
+    public static final String CODE_BLOCKED_KEY_PREFIX1 =
```
[the remark this answers is not in the document]
Comment (andrew-moores) [reply frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/LoginHandler.java]: Done

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/LoginHandler.java [hunk 2]
```diff
@@ -276,6 +267,58 @@ public APIGatewayProxyResponseEvent handleRequestWithUserContext(
...
+            if (configurationService.supportReauthSignoutEnabled() && isReauthJourney) {
+                codeStorageService.deleteIncorrectPasswordCountReauthJourney(request.getEmail());
+            } else {
+                blockUser(
+                        userProfile.getSubjectID(),
+                        request.getEmail(),
+                        updatedIncorrectPasswordCount,
+                        auditContext);
+            }
+            return generateApiGatewayProxyErrorResponse(400, ErrorResponse.ERROR_1028);
```
[the remark this answers is not in the document]
Comment (andrew-moores) [reply frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/LoginHandler.java:292]: Correct.  If reauth not enabled then the user will get logged off.

File: frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/LoginHandler.java [hunk 3]
```diff
@@ -276,6 +267,58 @@ public APIGatewayProxyResponseEvent handleRequestWithUserContext(
...
+        }
+
+        if (configurationService.supportReauthSignoutEnabled() && isReauthJourney) {
+            codeStorageService.increaseIncorrectPasswordCountReauthJourney(request.getEmail());
+        } else {
+            codeStorageService.increaseIncorrectPasswordCount(request.getEmail());
+        }
```
[the remark this answers is not in the document]
Comment (andrew-moores) [reply frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/LoginHandler.java:306]: Done.

**Review by mattvot — APPROVED** [review]

**Pull request merged**